modeling/11_Landsat_8_GTB_3class.Rmd

```diff
@@ -0,0 +1,365 @@
+
+## Apply model to image stack for Landsat
+
+### Load the image collection
+
+```{python}
+# filter stack for desired PRs
+ROWS = ee.List([27, 28])
+
+l8 = (ee.ImageCollection("LANDSAT/LC08/C02/T1_L2")
+  .filter(ee.Filter.lt('CLOUD_COVER', 80))
+  .filter(ee.Filter.eq('WRS_PATH', 26))
+  .filter(ee.Filter.inList('WRS_ROW', ROWS))
+  .filter(ee.Filter.gte('IMAGE_QUALITY_OLI', 7))
+  .filter(ee.Filter.calendarRange(4, 11, 'month'))
+  # mask high aerosol
+  .map(gf.mask_high_aerosol)
+  # mask pixels with any qa flags
+  .map(gf.mask_qa_flags)
+  # mask saturated pixels
+  .map(gf.apply_radsat_mask)
+  # apply scaling factors
+  .map(gf.applyScaleFactors))
+
+```
+
+### Load modeling AOIs and clip stack
+
+Note, some AOIs are too big to load in here and use as 'virtual' ee Feature
+Collections. Given that, we have manually uploaded the shapefiles as an Earth
+Engine Feature Collection.
+
+And then clip each image by that aoi
+
+```{python}
+l8_aoi = l8.map(gf.clip)
+```
+
+#### Helper functions
+
+```{python}
+# get CRS info
+img_crs = l8.first().projection()
+img_crsTrans = img_crs.getInfo().get('transform')
+```
+
+### consolidate stack by image date
+
+```{python}
+l8_aoi = l8_aoi.map(gf.addImageDate)
+
+# summarize by missionDate field
+uniqueMissDate_l8 = l8_aoi.aggregate_array('missDate').distinct()
+
+```
+
+### Create mosaics
+
+```{python}
+
+#function to apply the GTB model
+def applyGTB_ls8(image):
+  # Select the bands that correspond to the input features of the GTB model
+  imageFeatures = image.select(ls_input_feat)
+  missDate = image.get('missDate')
+  # Classify the image using the trained GTB model
+  classifiedImage = (imageFeatures
+    .classify(trainedGTB_ls8)
+    .set('missDate', missDate))
+  return image.addBands(classifiedImage)
+
+def applyPerMissionDate_ls8(missDate):
+  mission = ee.String(missDate).slice(0,9)
+  date = ee.String(missDate).slice(10,20)
+  short_stack = (l8
+    .filter(ee.Filter.eq('SPACECRAFT_ID', mission))
+    .filter(ee.Filter.eq('DATE_ACQUIRED', date)))
+  oneMissDate = short_stack.mean()
+  ls_miss_date_GTB = applyGTB_ls8(oneMissDate)
+  ls_GTB_class = gf.extract_3classes(ls_miss_date_GTB)
+  return (ls_GTB_class.set('missDate', missDate))
+
+def mosaicStack_l8(missDate):
+  md_GTB = applyPerMissionDate_ls8(missDate)
+  return md_GTB
+
+newStack_list_l8 = uniqueMissDate_l8.map(mosaicStack_l8)
+newStack_l8 = ee.ImageCollection(newStack_list_l8)
+
+```
+
+### Lighten up each of the stacks to only the bands we care about
+
+```{python}
+lightStack_l8 = newStack_l8.select([
+    'classified',
+    'cloud',
+    'openWater',
+    'sediment'
+    ])
+
+```
+
+
+## Export GeoTiffs to drive
+
+### GTB images for Landsat 8
+
+```{python}
+aoi_ee = ee.FeatureCollection('projects/ee-ross-superior/assets/aoi/Superior_AOI_modeling')
+
+def clip(image):
+  return image.clip(aoi_ee.geometry())
+
+
+date_length_8 = len(uniqueMissDate_l8.getInfo())
+
+# export tif to drive
+for d in range(63, date_length_8):
```

Comment:
What's the reason for '63' here?

Reply:
Ha! This is because my loop stopped at index 63, so I had to restart it there.